### PR TITLE
Manage orientation of detector

### LINKED ISSUE #1994: Poni file generated by Dioptas are incompatible with pyFAI ?
Apparently the images are flipped somewhere.

Spotted at ID27
Thanks @t20100 @loichuder @EdgarGF93
attn @CPrescher

### PR DESCRIPTION
**Pull request by kif**
This is likely to be a beefy one ...

* [X] create the detector orientation tag: value 0 to be used for "do not care" else value [1..8]
* [x] Use this tag when calculating the pixel center coordinate
* [x] Use this tag when calculating the pixel corner coordinate
* [x] Use this tag when calculating the x,y,z of a list of pixels (used while calibrating)
* [X] Save / restore this attribute into the Nexus detector file
* [X] Save / restore into the PONI-file
* [X] Save / restore into the config
* [x] Add some tests specific to orientation change
* [ ] ~~Ensure~~ the Detector cache gets reset when changing the orientation (--> use a property)
* [ ] ~~Ensure~~ the geometry cache gets reset when changing the orientation ( what is the best way of doing this ?)

close #1994

### CODE REVIEW

File: src/pyFAI/detectors/orientation.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    FlipHorizontal = 2
+    Rotate180 = 3
+    FlipVertical = 4
+    Transpose = 5
+    Rotate270 = 6
+    Transverse = 7
+    Rotate90 = 8
+
+Orientation(0).__doc__ = "An orientation flag is not set."
+Orientation(1).__doc__ = "Camera default. Origin at the top left of the image when looking at the sample."
```
Comment (loichuder) [inline src/pyFAI/detectors/orientation.py:61]: So origin at top right when looking from the sample ?

Comment (kif) [reply src/pyFAI/detectors/orientation.py:61]: This enum is the default one used in EXIF, for visible light cameras.

In pyFAI, the origin is at the bottom left when looking from the sample. Thus the orientation should be flipped-ud for the origin at the bottom and flipped-rl to change the direction of sight. This makes the default orientation used in pyFAI to `3` or `Rotate180`.

I just got the answer from Dectris that their detector have to be seen from the sample and not from the other side.

At this stage, orientation is just shipped around and used nowhere ...

Comment (loichuder) [reply src/pyFAI/detectors/orientation.py:61]: Ok.

When speaking about images, looking at the detector from the sample makes sense. Looking at the sample from the detector is a bit harder to figure out.

Comment (kif) [reply src/pyFAI/detectors/orientation.py:61]: Looking at the sample is the standard when considering camera ... outside the X-ray scattering world.

https://www.researchgate.net/profile/Romain-Neuville/publication/330564129/figure/fig2/AS:718599043952648@1548338771876/The-virtual-camera-settings-camera-position-camera-orientation-focal-length-and.jpg

Comment (kif) [reply src/pyFAI/detectors/orientation.py:61]: https://www.researchgate.net/profile/Sunil-Kumar-Kopparapu/publication/220632389/figure/fig2/AS:883184350789637@1587578966909/Orientation-angles-rotation-about-the-camera-x-y-z-axes-are-referred-to-as-pitch.pbm

Comment (loichuder) [reply src/pyFAI/detectors/orientation.py:61]: Yeah. After thinking again about it this morning, this is indeed just a personal feeling.

**Comment by kif** [comment]
I believe this PR is beefy enough -> ready for merge (once test pass)
Related issues are linked in the project

**Comment by kif** [comment]
Checked with Edgar that flipped image can be processed as expected. "just" declare an orientation different from "3"

File: src/pyFAI/detectors/_common.py [hunk 2]
```diff
@@ -549,6 +566,31 @@ def setFit2D(self, **kwarg):
             elif kw == "splineFile":
                 self.set_splineFile(kwarg[kw])
 
+    def _calc_pixel_index_from_orientation(self, plus_one=False):
+        """Calculate the pixel index when considereing the different orientations"""
```
Comment (loichuder) [inline src/pyFAI/detectors/_common.py:570]:
```suggestion
        """Calculate the pixel index when considering the different orientations"""
```

File: src/pyFAI/detectors/_dectris.py [hunk 3]
```diff
@@ -438,6 +459,8 @@ def __init__(self, pixel1=172e-6, pixel2=172e-6, max_shape=None, module_size=Non
 
     def __repr__(self):
         txt = f"Detector {self.name}\t PixelSize= {self._pixel1:.3e}, {self._pixel2:.3e} m"
+        if self.orientation>1:
```
Comment (loichuder) [inline src/pyFAI/detectors/_dectris.py:462]:
```suggestion
        if self.orientation > 0:
```

I guess the orientation info should be displayed also for `orientation = 1`.

File: src/pyFAI/detectors/_dectris.py [hunk 4]
```diff
@@ -742,12 +780,14 @@ class Pilatus4(_Dectris):
...
-    def __init__(self, pixel1=150e-6, pixel2=150e-6, max_shape=None):
-        super(Pilatus4, self).__init__(pixel1=pixel1, pixel2=pixel2, max_shape=max_shape)
+    def __init__(self, pixel1=150e-6, pixel2=150e-6, max_shape=None, orientation=0):
+        super(Pilatus4, self).__init__(pixel1=pixel1, pixel2=pixel2, max_shape=max_shape, orientation=orientation)
         self.module_size = tuple(self.MODULE_SIZE)
 
     def __repr__(self):
-        return f"Detector {self.name}\t PixelSize= {self._pixel1:.3e}, {self._pixel2:.3e} m"
+        txt = f"Detector {self.name}\t PixelSize= {self._pixel1:.3e}, {self._pixel2:.3e} m"
+        if self.orientation>1:
```
Comment (loichuder) [inline src/pyFAI/detectors/_dectris.py:789]:
```suggestion
        if self.orientation > 0:
```

File: src/pyFAI/detectors/_rayonix.py [hunk 5]
```diff
@@ -640,4 +643,5 @@ def set_config(self, config):
                 raise err
         self.set_pixel1(config.get("pixel1"))
         self.set_pixel2(config.get("pixel2"))
+        self.orientation = Orientation(config.get("orientatio", 3))
```
Comment (loichuder) [inline src/pyFAI/detectors/_rayonix.py:646]:
```suggestion
        self.orientation = Orientation(config.get("orientation", 3))
```

File: src/pyFAI/detectors/orientation.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+Orientation(2).__doc__ = "Origin at the top left of the image when looking from the sample."
+Orientation(3).__doc__ = "Native orientation of pyFAI. Origin at the bottom left when looking from the sample"
```
Comment (loichuder) [inline src/pyFAI/detectors/orientation.py:63]:
```suggestion
Orientation(3).__doc__ = "Native orientation of pyFAI. Origin at the bottom left when looking from the sample."
```

File: src/pyFAI/test/test_geometry.py [hunk 7]
```diff
@@ -348,6 +361,8 @@ def test_corner_array(self):
         for data, space in params:
             with self.subTest(data=data, space=space):
                 geo = geometry.Geometry(**data)
+                print(data)
+                print(geo.detector.orientation)
```
Comment (loichuder) [inline src/pyFAI/test/test_geometry.py:365]: To be removed, I guess.

File: src/pyFAI/test/test_detector.py [hunk 8]
```diff
@@ -378,11 +379,96 @@ def test_abstract(self):
...
+
+        #orient3
+        r1, r2, _ = self.orient3.calc_cartesian_positions(Y, X)
+        ref1, ref2, _ = self.orient3.calc_cartesian_positions()
+        p1 = ref1[Y, X]
+        p2 = ref2[Y, X]
+        self.assertTrue(numpy.allclose(r1, p1), "orient 3 dim1,y points")
+        self.assertTrue(numpy.allclose(r2, p2), "orient 3 dim2,x points")
+
+        #orient1
```
Comment (loichuder) [inline src/pyFAI/test/test_detector.py:458]:
```suggestion
        #orient4
```

File: src/pyFAI/test/test_detector.py [hunk 9]
```diff
@@ -378,11 +379,96 @@ def test_abstract(self):
...
+        p1, p2, _ = self.orient1.calc_cartesian_positions(center=False)
+        #orient2 -> flip rl
+        r1, r2, _ = self.orient2.calc_cartesian_positions(center=False)
+        self.assertTrue(numpy.allclose(p1, numpy.fliplr(r1)), "orient 2vs1 dim1,y corner")
+        self.assertTrue(numpy.allclose(p2, numpy.fliplr(r2)), "orient 2vs1 dim2,x corner")
+        #orient3 -< rotate180
+        r1, r2, _ = self.orient3.calc_cartesian_positions(center=False)
+        self.assertTrue(numpy.allclose(p1, numpy.flipud(numpy.fliplr(r1))), "orient 3vs1 dim1,y corner")
+        self.assertTrue(numpy.allclose(p2, numpy.flipud(numpy.fliplr(r2))), "orient 3vs1 dim2,x corner")
+        #orient3 -> flip u-d
```
Comment (loichuder) [inline src/pyFAI/test/test_detector.py:424]:
```suggestion
        #orient4 -> flip u-d
```

File: src/pyFAI/test/test_detector.py [hunk 10]
```diff
@@ -378,11 +379,96 @@ def test_abstract(self):
...
+        p1, p2, _ = self.orient1.calc_cartesian_positions()
+        #orient2 -> flip rl
+        r1, r2, _ = self.orient2.calc_cartesian_positions()
+        self.assertTrue(numpy.allclose(p1, numpy.fliplr(r1)), "orient 2vs1 dim1,y center")
+        self.assertTrue(numpy.allclose(p2, numpy.fliplr(r2)), "orient 2vs1 dim2,x center")
+        #orient3 -< rotate180
+        r1, r2, _ = self.orient3.calc_cartesian_positions()
+        self.assertTrue(numpy.allclose(p1, numpy.flipud(numpy.fliplr(r1))), "orient 3vs1 dim1,y center")
+        self.assertTrue(numpy.allclose(p2, numpy.flipud(numpy.fliplr(r2))), "orient 3vs1 dim2,x center")
+        #orient3 -> flip u-d
```
Comment (loichuder) [inline src/pyFAI/test/test_detector.py:409]:
```suggestion
        #orient4 -> flip u-d
```

**Comment by EdgarGF93** [comment]
It does fix the dioptas incompatibility.
I did the calibration with dioptas and pyFAI on the same .edf file, from a Pilatus1M.
As we said, the PONI and rotation values generated by pyFAI-calib2 and dioptas are different, therefore the result of AzimuthalIntegrator.integrate2d is wrong for the dioptas file. After resetting the azimuthal integrator instance and changing manually the orientation of the detector instance to '2', we get a good reshaping of the pattern (with the azimuthal axis inverted):



image

File: src/pyFAI/detectors/_rayonix.py [hunk 11]
```diff
@@ -595,6 +596,7 @@ def set_config(self, config):
                 raise err
         self.set_pixel1(config.get("pixel1"))
         self.set_pixel2(config.get("pixel2"))
+        self.oientation = Orientation(config.get("orientation", 3))
```
Comment (EdgarGF93) [inline src/pyFAI/detectors/_rayonix.py:599]:
```suggestion
        self.orientation = Orientation(config.get("orientation", 3))
```

**Review by EdgarGF93 — CHANGES_REQUESTED** [review]

File: src/pyFAI/detectors/orientation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kif) [reply src/pyFAI/detectors/orientation.py]: well spotted

**Comment by t20100** [comment]
attn @CPrescher, a "detector orientation tag" has been added to pyFAI to deal with image flip (not released yet).

**Pull request merged**